File: WEEK7/Js/bj7_11049.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const matrix = [];
for (let i = 1; i <= n; i++) {
    matrix.push(input[i].split(' ').map(Number));
}

const dp = Array.from({ length: n }, () => new Int32Array(n).fill(0));

for (let gap = 1; gap < n; gap++) {
    for (let i = 0; i + gap < n; i++) {
        const j = i + gap;
        dp[i][j] = Infinity;
        for (let k = i; k < j; k++) {
            const cost = dp[i][k] + dp[k + 1][j] + matrix[i][0] * matrix[k][1] * matrix[j][1];
            dp[i][j] = Math.min(dp[i][j], cost);
        }
    }
}

console.log(dp[0][n - 1]);

File: WEEK7/Python/bj7_11049.py
import sys

n = int(sys.stdin.readline())
matrix = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
dp = [[0] * n for _ in range(n)]

for gap in range(1, n):
    for i in range(n - gap):
        j = i + gap
        dp[i][j] = float("inf")
        for k in range(i, j):
            cost = dp[i][k] + dp[k + 1][j] + matrix[i][0] * matrix[k][1] * matrix[j][1]
            dp[i][j] = min(dp[i][j], cost)

print(dp[0][n - 1])
